hypothesis-python/tests/cover/test_slices.py

```diff
@@ -17,40 +17,57 @@
 
 from __future__ import absolute_import, division, print_function
 
+import pytest
+
 from hypothesis import given, settings, strategies as st
 from tests.common.debug import assert_all_examples, find_any, minimal
 
+use_several_sizes = pytest.mark.parametrize("size", [1, 2, 5, 10, 100, 1000])
+
 
-def test_stop_stays_within_bounds():
-    size = 1000
+@use_several_sizes
+def test_stop_stays_within_bounds(size):
     assert_all_examples(
         st.slices(size), lambda x: x.stop is None or (x.stop >= 0 and x.stop <= size)
     )
 
 
-def test_start_stay_within_bounds():
-    size = 1000
+@use_several_sizes
+def test_start_stay_within_bounds(size):
     assert_all_examples(
-        st.slices(size),
-        lambda x: x.start is None or (x.start >= 0 and x.start <= size - 1),
+        st.slices(size), lambda x: x.start is None or (x.start >= 0 and x.start <= size)
     )
 
 
-def test_step_stays_within_bounds():
-    size = 1000
+@use_several_sizes
+def test_step_stays_within_bounds(size):
     # indices -> (start, stop, step)
+    # Stop is exclusive so we use -1 as the bottom
     assert_all_examples(
         st.slices(size),
-        lambda x: x.indices(size)[0] + x.indices(size)[2] <= size
-        and x.indices(size)[0] + x.indices(size)[2] >= 0,
+        lambda x: (
+            (
+                x.indices(size)[0] + x.indices(size)[2] <= size
```

Comment:
`x.start + x.step <= size` would be clearer to me here.  You can also remove the inner set of parentheses, which aren't doing anything.  (Just remember to run Black afterwards!)

Reply:
The start and step behave a bit differently since they can be 'None'. splice translates that automatically to indices which makes a bit easier (in my opinion) to figure out what's happening. To further complicate things None can be translated differently depending on start and step.

Reply:
Oh, right!  That would be worth a comment then :smile: